### PR TITLE
OAS 3.1 for JSON Schemas

### PR DESCRIPTION
**Pull request by Dashron**
## 🧰 Changes

This is all the work necessary to support OAS 3.1 for JSON schemas.

pre OAS 3.1 your JSON schema could only be draft `04`. 3.1 your JSON schema can be anything, as determined by the `$schema` value in your schema, the value of `jsonSchemaDialect` or the default, draft `2020-12`.

These updates to the OAS library now ensure that the schemas you are working with always have the accurate `$schema` value, following the previously mentioned chain of priorities.

Along the way I converted the following to typescript
- operation
- get-parameters-as-json-schema

## 🧬 QA & Testing

- tests

### CODE REVIEW

**Comment by Dashron** [comment]
There's some *really* ugly typing going on in here, but let's get the typescript conversion over before we do code cleanup. I gave code cleanup a shot and it was *really* problematic.

File: src/operation/get-response-as-json-schema.ts [hunk 1]
```diff
@@ -11,27 +16,38 @@ const { json: isJSON } = require('../lib/matches-mimetype').default;
...
 
-  const headersSchema = {
+  const headersSchema: SchemaObject = {
     type: 'object',
     properties: {},
   };
 
   Object.keys(headers).forEach(key => {
-    if (headers[key] && headers[key].schema) {
+    if (headers[key] && (headers[key] as OpenAPIV3.HeaderObject | OpenAPIV3_1.HeaderObject).schema) {
```
Comment (erunion) [inline src/operation/get-response-as-json-schema.ts:28]: Might want to make a `HeaderObject` in `rmoas.types` that shortcuts having to repeat `OpenAPIV3.HeaderObject | OpenAPIV3_1.HeaderObject`.

**Review by erunion — CHANGES_REQUESTED** [review]

**Comment by erunion** [comment]
I also made a couple changes for minor things I found, you can see all my changes here: https://github.com/readmeio/oas/compare//58d150e746f52e9a25fdfa4d27424ebce94b412d...1841ce8?expand=1

* Pulled back some code docs that got deleted
* Moved `import type` declarations to the top of the files they're in so we're consistent
* Moved the `ResponseExamples` and `RequestBodyExamples` exports back to using the exports from `get-*-examples.ts` because they're already there. Also removed the `| false` declaration that got added because those arrays of examples will never have a falsy entry because we do `.filter(Boolean)` on it.

File: __tests__/operation.test.ts [hunk 2]
```diff
@@ -12,13 +13,13 @@ describe('#constructor', () => {
   it('should accept an Oas instance into a definition to be used', () => {
     const operation = new Operation(oas, '/test', 'get', { summary: 'operation summary' });
     expect(operation.schema).toStrictEqual({ summary: 'operation summary' });
-    expect(operation.api).toStrictEqual(petstore);
+    expect(operation['api']).toStrictEqual(petstore);
```
[the remark this answers is not in the document]
Comment (Dashron) [reply __tests__/operation.test.ts:16]: At one point `api` was private, and this bypasses typescript's public/private check. It looks like it's no longer private though so it works with dot notation.

[the remark this answers is not in the document]
Comment (erunion) [reply __tests__/operation.test.ts:16]: I think there's also a `getApi()` accessor on `Operation` if you want to use that instead.

File: src/lib/openapi-to-json-schema.ts [hunk 3]
```diff
@@ -467,8 +507,9 @@ function toJSONSchema(data, opts = {}) {
   }
 
   // Enums should not have duplicated items as those will break AJV validation.
-  if ('enum' in schema && Array.isArray(schema.enum)) {
-    schema.enum = [...new Set(schema.enum)];
+  if (RMOAS.isSchema(schema, isPolymorphicAllOfChild) && 'enum' in schema && Array.isArray(schema.enum)) {
+    // If we ever target ES6 for typescript we can drop this array.from. https://stackoverflow.com/questions/33464504/using-spread-syntax-and-new-set-with-typescript/56870548
+    schema.enum = [...Array.from(new Set(schema.enum))];
```
[the remark this answers is not in the document]
Comment (Dashron) [reply src/lib/openapi-to-json-schema.ts:512]: weird, that all came from fixing some typing issues, looks to still be fine typewise without it.

File: src/operation.ts [hunk 4]
```diff
@@ -320,25 +321,29 @@ export default class Operation {
...
+    const oasTagMap: Map<string, RMOAS.TagObject> = new Map();
     if ('tags' in this.api) {
-      this.api.tags.forEach(tag => {
-        oasTags.set(tag.name, tag);
+      this.api.tags.forEach((tag: RMOAS.TagObject) => {
+        oasTagMap.set(tag.name, tag);
       });
     }
 
+    const oasTags = Object.fromEntries(oasTagMap);
```
[the remark this answers is not in the document]
Comment (Dashron) [reply src/operation.ts:342]: I don't remember specifically, but I'm sure it's typing related. Happy to mess around with restoring the old version tomorrow morning if you'd like.

**Comment by Dashron** [comment]
Just pushed up fixes for everything. LMK what you think.

**Review by erunion — APPROVED** [review]

**Pull request merged**